tkNtuple/WriteMemoryFile.cc

```diff
@@ -860,6 +875,13 @@ int main(int argc, char *argv[])
       int ilay = -1;
       ilay = lnc.convertLayerNumber(pix_det->at(ipix),pix_lay->at(ipix),useMatched,-1,pix_z->at(ipix)>0);
       if (ilay<0) continue;
+
+      unsigned int imoduleid;
+      {
+        auto ii = module_shortId_hash[ilay].emplace(pix_detId->at(ipix), (unsigned int) module_shortId_hash[ilay].size());
+        imoduleid = ii.first->second;
```

Comment:
(For CMSSW side) Is it enough to provide a unique 12-bit number for each module on a layer? Or does the "short id" as implemented here carry any other meaning?

(I'm just thinking that for CMSSW it would be more natural to derive this kind of mapping from DetId to a "short id" once from the geometry instead of loops over hits)

Reply:
Yes, it would be preferable to extract it from DetId and store that. I wasn't exactly sure how this changes with different tracker phases so I picked a lazy way out. We also have more bits if we need them.

Also, don't forget that we use stereo/mono as different layers.

Do you know how to do it in a reasonable way?

Reply:
(In lossyspeak) I was thinking to just take the geometry, ask for all modules, loop over them and use a similar `module_shortId_has_` trick to map DetId to a "sequence number".

By very quick look the `GeomDet`s actually have a global indexing 
https://github.com/cms-sw/cmssw/blob/43c5f0689f3839d0e430dcf63cbf78c2ece55adc/Geometry/CommonDetUnit/interface/GeomDet.h#L86-L88
that is used at least for tracker
https://github.com/cms-sw/cmssw/blob/43c5f0689f3839d0e430dcf63cbf78c2ece55adc/Geometry/TrackerGeometryBuilder/src/TrackerGeometry.cc#L141
One option would be to use that directly (the geometry seems to internally map DetIds to `const GeomDet*` with a `std::unordered_map` so it should not be much slower than going through `std::unordered_map<unsigned, unsigned>`). I'm not sure if everything would fit in 12 bits though (in the available 18 for sure, I don't know on the top of my head whether  phase2 tracker could hit 260k modules in total). In a sense this index would be overkill (it would be unique through the tracker), but it would still be unique within a layer :)

Reply:
Slava once pointed me to this: https://github.com/cms-sw/cmssw/blob/02d4198c0b6615287fd88e9a8ff650aea994412e/Geometry/TrackerNumberingBuilder/README.md

And it scared me enough to go the hash_map way :) The largest number of modules I saw for my scheme was between 860 and 880 ... for the outer TOB layer.

But you are right, we can have an id that is unique across more than just "our" layer.

On the other hand, we could also have an id, that is only unique across potentially overlapping modules ... but this then becomes a "map coloring problem" and is probably even harder.

Now, if we ever need to access per module state (orientation, dead/always-on strips), this id would also be a "natural" index into the module_state vector.

Sigh, I guess I'm not helping. :)